### PR TITLE
identity: Make identity allocations observable

### PR DESCRIPTION
**Pull request by mhofstetter**
This makes the identity allocator changes observable in order to provide cells backend-agnostic access to identity allocations.

The identity allocation changes are now provided in the hive as stream.Observable[cache.IdentityChange]. When observing the initial listing is first waited for, then the current state is replayed followed by sync event and updates to state:

`[ (wait for OnListDone()), Upsert, Upsert, Sync, Upsert, Delete, ... ]`

Related to https://github.com/cilium/cilium/issues/25898 as this lays the foundation to replace the use of `Resource[CiliumIdentity]` with the observable identity allocator. This comes with the advantage of getting rid of the duplicated K8s Watcher for CiliumIdentities and the additional support for KVStore identity backend. A [follow up PR](https://github.com/cilium/cilium/pull/26375) will refactor the authentication related part (garbage collection).

Original PR by @joamaki: https://github.com/cilium/cilium/pull/26229

### CODE REVIEW

File: pkg/allocator/cache.go [hunk 1]
```diff
@@ -274,3 +289,43 @@ func (c *cache) numEntries() int {
...
+	Key  AllocatorKey
+}
+
+// Observe the allocator changes. Conforms to stream.Observable.
+// Replays the current state of the cache when subscribing.
+func (c *cache) Observe(ctx context.Context, next func(AllocatorChange), complete func(error)) {
+	go func() {
+		// Wait until initial listing has completed before
+		// replaying the state.
+		<-c.listDone
```
Comment (giorio94) [inline pkg/allocator/cache.go]: I'd change this into a select statement which also checks for `ctx.Done()` (I guess that in that case `complete()` should also be called).

File: pkg/allocator/cache.go [hunk 2]
```diff
@@ -58,15 +59,21 @@ type cache struct {
 	// watcher is started with the conditions marked as done when the
 	// watcher has exited
 	stopWatchWg sync.WaitGroup
+
+	changeSrc         stream.Observable[AllocatorChange]
+	emitChange        func(AllocatorChange)
```
Comment (giorio94) [inline pkg/allocator/cache.go:64]: Should this be `emitChangeSrc` for consistency with the other variables?

File: pkg/allocator/cache.go [hunk 3]
```diff
@@ -58,15 +59,21 @@ type cache struct {
...
+	completeChangeSrc func(error)
```
Comment (giorio94) [inline pkg/allocator/cache.go:65]: Should this be called when `cache.stop()` is executed? It appears that `start` cannot be called again afterwards, because the `stopChan` channel is already closed.

File: pkg/allocator/allocator_test.go [hunk 4]
```diff
@@ -363,6 +365,65 @@ func (s *AllocatorSuite) TestAllocateCached(c *C) {
 	testAllocator(c, idpool.ID(256), randomTestName(), "a") // enable use of local cache
 }
 
+func (s *AllocatorSuite) TestObserveAllocatorChanges(c *C) {
```
Comment (giorio94) [inline pkg/allocator/allocator_test.go]: Nit. AFAIU, new tests should leverage the testing framework rather than checkmate.

File: pkg/allocator/cache.go [hunk 5]
```diff
@@ -170,6 +183,8 @@ func (c *cache) onDeleteLocked(id idpool.ID, key AllocatorKey) {
 	delete(c.nextCache, id)
 	a.idPool.Insert(id)
 
+	c.emitChange(AllocatorChange{Kind: AllocatorChangeDelete, ID: id, Key: key})
```
Comment (giorio94) [inline pkg/allocator/cache.go:184]: This comment is actually unrelated from this line. A new key can be added into the key cache also through the `insert` function below, and a change will not be emitted in that case. Yet, when an observer is started and the current state is replayed, that event would be emitted, which is inconsistent. 

Currently, an event is not sent to the events channel when `insert` is called, which might also need to be considered if planning to replace it with the observable pattern.

File: pkg/allocator/cache.go [hunk 6]
```diff
@@ -122,6 +129,10 @@ func (c *cache) OnAdd(id idpool.ID, key AllocatorKey) {
 	}
 	c.allocator.idPool.Remove(id)
 
+	c.emitChange(AllocatorChange{Kind: AllocatorChangeUpsert, ID: id, Key: key})
+
+	// TODO: Remove sendEvent in favour of the observable. Might need
+	// to support modify? Or just Upsert?
```
Comment (giorio94) [inline pkg/allocator/cache.go]: Just to double-check., is this supposed to be tackled in a followup PR? Even in this case, it might be worth to decide whether to split the UpsertEvent into create and update to avoid the need for a subsequent refactoring (e.g., [here](https://github.com/cilium/cilium/blob/main/pkg/identity/cache/cache.go#L153-L160) it seems that update events are ignored).

**Review by giorio94 — COMMENTED** [review]
Nice! I've left a few comments inline.

File: pkg/allocator/cache.go (hunk 1 above)
Comment (mhofstetter) [reply pkg/allocator/cache.go]: agree. refactored towards the select statement calling `complete & returning` in case of `ctx.Done`

File: pkg/allocator/cache.go (hunk 2 above)
Comment (mhofstetter) [reply pkg/allocator/cache.go:64]: i would keep it as is. `emitChange` makes no reference to the source (_emitting a change_), whereas `changeSrc` _is the actual change source_ and `completeChangeSrc` _completes the change source_.

but this is just my opinion of how i read this.

File: pkg/allocator/cache.go (hunk 3 above)
Comment (mhofstetter) [reply pkg/allocator/cache.go:65]: good idea. changed

File: pkg/allocator/allocator_test.go (hunk 4 above)
Comment (mhofstetter) [reply pkg/allocator/allocator_test.go]: thanks. refactored towards go testing framework in combination with testify

File: pkg/allocator/cache.go (hunk 2 above)
Comment (giorio94) [reply pkg/allocator/cache.go:64]: Sounds fine to me :+1:

File: pkg/allocator/cache.go (hunk 5 above)
Comment (mhofstetter) [reply pkg/allocator/cache.go:184]: > This comment is actually unrelated from this line. A new key can be added into the key cache also through the `insert` function below, and a change will not be emitted in that case. Yet, when an observer is started and the current state is replayed, that event would be emitted, which is inconsistent.
> 
> Currently, an event is not sent to the events channel when `insert` is called, which might also need to be considered if planning to replace it with the observable pattern.

having a quick glance, it looks like `insert` gets called by `Allocate` which calls out to the backend via `lockedAllocate`. Eventually, after allocating the ID, the backend calls the cache via `OnAdd` (maybe `OnModify`) which triggers the observer. :thinking: But i didn't take a look at the branch with `localKeys`.

> Currently, an event is not sent to the events channel when insert is called, which might also need to be considered if planning to replace it with the observable pattern.

Above statement might also be the reason why no event is sent in case of an `insert`.

File: pkg/allocator/cache.go (hunk 6 above)
Comment (mhofstetter) [reply pkg/allocator/cache.go]: AFAIK - nothing planned so far. But anyway - would be easy to refactor into any direction as long as there aren't many usages of this observable. `Upsert` makes it a 1:1 replacement when replacing `Resource[CiliumIdentity]`.

**Comment by mhofstetter** [comment]
@giorio94 thanks for your fast feedback :rocket: i applied your suggestions - PTAL!

File: pkg/allocator/cache.go (hunk 5 above)
Comment (giorio94) [reply pkg/allocator/cache.go:184]: Yeah, I agree that likely it is not an issue because the event will be eventually emitted when the creation event is received from the kvstore, although I'm not an expert of this code path. Let's also see the Chris' feedback.

File: pkg/allocator/cache.go (hunk 6 above)
Comment (giorio94) [reply pkg/allocator/cache.go]: I see. The typical issue is only that then usages pile up and refactoring becomes more onerous, but I agree that it is not something that needs to be addressed right now.

**Review by giorio94 — APPROVED** [review]
/lgtm

File: pkg/identity/cache/allocator.go [hunk 7]
```diff
@@ -482,3 +486,43 @@ func (m *CachingIdentityAllocator) RemoveRemoteIdentities(name string) {
...
+const (
+	IdentityChangeSync   IdentityChangeKind = IdentityChangeKind(allocator.AllocatorChangeSync)
+	IdentityChangeUpsert IdentityChangeKind = IdentityChangeKind(allocator.AllocatorChangeUpsert)
+	IdentityChangeDelete IdentityChangeKind = IdentityChangeKind(allocator.AllocatorChangeDelete)
+)
+
+type IdentityChange struct {
+	Kind   IdentityChangeKind
+	ID     identity.NumericIdentity
+	Labels map[string]string
```
Comment (joestringer) [inline pkg/identity/cache/allocator.go]: Any particular reason not to use the native types for this?

```suggestion
	Labels labels.Labels
```

File: pkg/identity/cache/allocator.go [hunk 8]
```diff
@@ -482,3 +486,43 @@ func (m *CachingIdentityAllocator) RemoveRemoteIdentities(name string) {
...
+				next(IdentityChange{
+					Kind:   IdentityChangeKind(change.Kind),
+					ID:     identity.NumericIdentity(change.ID),
+					Labels: labels,
+				})
+			},
+			complete,
+		)
+	} else {
+		go complete(nil)
```
Comment (joestringer) [inline pkg/identity/cache/allocator.go]: Is there any documentation around the goroutine handling for complete functions? It's a bit hard for me to understand the implications of asynchronously running this generic function.

File: pkg/allocator/cache.go [hunk 9]
```diff
@@ -274,3 +290,48 @@ func (c *cache) numEntries() int {
...
+type AllocatorChange struct {
+	Kind AllocatorChangeKind
+	ID   idpool.ID
+	Key  AllocatorKey
+}
+
+// Observe the allocator changes. Conforms to stream.Observable.
+// Replays the current state of the cache when subscribing.
+func (c *cache) Observe(ctx context.Context, next func(AllocatorChange), complete func(error)) {
+	go func() {
```
Comment (joestringer) [inline pkg/allocator/cache.go:312]: What's the lifecycle of these goroutines?

Comment (joestringer) [reply pkg/allocator/cache.go:312]: I partly see the motivation here given that there's the blocking call on `listDone` below, I guess each time a stream observer subscribes to events of this type they'll get a goroutine for their own stream.

File: pkg/allocator/cache.go [hunk 10]
```diff
@@ -274,3 +290,48 @@ func (c *cache) numEntries() int {
...
+		// replaying the state.
+		select {
+		case <-c.listDone:
+		case <-ctx.Done():
+			complete(ctx.Err())
+			return
+		}
+
+		c.mutex.RLock()
+		defer c.mutex.RUnlock()
```
Comment (joestringer) [inline pkg/allocator/cache.go:323]: When exactly does this get released and how does that relate to the overall stream lifecycle?

**Review by joestringer — COMMENTED** [review]
Some minor comments and concerns around type reuse and lifecycle questions. I'm clearly not yet qualified enough to review code that uses streams.

File: pkg/identity/cache/allocator.go (hunk 7 above)
Comment (mhofstetter) [reply pkg/identity/cache/allocator.go]: Not that i'm aware of. Thanks for the input - I refactored it to use `labels.Labels`

File: pkg/identity/cache/allocator.go (hunk 8 above)
Comment (mhofstetter) [reply pkg/identity/cache/allocator.go]: `Observable.Observe` is documented as followed.

```
// Observable implementations are allowed to call 'next' and 'complete'
// from any goroutine, but never concurrently.
Observe(ctx context.Context, next func(T), complete func(error))
```

https://github.com/cilium/cilium/blob/main/pkg/stream/observable.go#L22-L24

Hence the caller should be aware that it gets called asynchronously.

File: pkg/allocator/cache.go (hunk 9 above)
Comment (mhofstetter) [reply pkg/allocator/cache.go:312]: That's exactly how i see it too.

File: pkg/allocator/cache.go (hunk 10 above)
Comment (mhofstetter) [reply pkg/allocator/cache.go:323]: It will get released as soon as the observation starts -> `c.changeSrc.Observe` won't block because it will start a new goroutine in any case - even if it's just to call `complete`.

**Comment by mhofstetter** [comment]
> Some minor comments and concerns around type reuse and lifecycle questions. I'm clearly not yet qualified enough to review code that uses streams.

@joestringer thanks for your feedback and questions. I included your feedback (`labels.Labels`) in the latest force push and tried to answer your questions as good as possible :pray:

File: pkg/identity/cache/allocator.go (hunk 8 above)
Comment (jrajahalme) [reply pkg/identity/cache/allocator.go]: Is it possible that `Observe` gets first called with a `nil` `m.IdentityAllocator` (it spawns a go-routine to call `complete()`), and then later gets called again with the same `next` and `complete` and with non-`nil` `m.IdentityAllocator` when it will then call `next()`? If so, it could happen that `next()` and `complete()` are running concurrently?

Comment (jrajahalme) [reply pkg/identity/cache/allocator.go]: In other words, is it allowed for the `complete()` to be running after `Observe(), ctx, next, complete)` has returned?

Comment (mhofstetter) [reply pkg/identity/cache/allocator.go]: @jrajahalme  As outlined by the API documentation, it only guarantees this (never calling Next & Complete concurrently) for one call to `Observe`.

It's not guaranteed when registering multiple times with the same next/complete to the same observable - but IMO that doesn't make any sense anyway :)

File: pkg/allocator/cache.go (hunk 9 above)
Comment (joestringer) [reply pkg/allocator/cache.go:312]: Right, but how often are goroutines here spawned and how do we know they're cleaned up?

Comment (mhofstetter) [reply pkg/allocator/cache.go:312]: Like for most other usages of an Observable, a goroutine gets spawned once per registration to the observable (e.g. once by the auth gc at Cilium startup (as a first usage)). Typically the goroutine of the Observable gets terminated once the passed `context.Context` has been cancelled or the Observables source marks the Observable as `completed`. Therefore, most of the more complex Observables within Cilium use `stream.Multicast` on the providing side which provides this facility. This is also the case for the IdentityAllocatorCache (pkg/allocator/allocator.go).

And at best, on the consuming side, the new [hive jobs framework](https://docs.cilium.io/en/latest/contributing/development/hive/#job-groups) (or something similar) is used to provide a unified way to consume an Observable and hook it to a handler.

`jobGroup.Add(job.Observer("auth identities gc events", mapGC.handleIdentityChange, params.IdentityChanges))` 

(with `params.IdentityChanges` being the Observable) It's done this way in the auth related follow up PR (https://github.com/cilium/cilium/pull/26375)

This approach takes care of properly hooking to an Observable, propagating the context (which is attached to the hive lifecycle) and therefore cleaning up the related goroutines once Cilium shuts down (if not completed before).

---

One special thing in this use-case is the additional goroutine in `pkg/allocator/cache.go`. It ensures that the current state of the cache gets re-played at the beginning - to mimic the behaviour of `Resource[CiliumIdentity]`. But this is a short-lived goroutine which should terminate as soon as the state of the cache has been re-played (or the context has been cancelled in the meantime) and the observation got delegated to the actual Observable mentioned above.

**Comment by mhofstetter** [comment]
Added @dylandreimerink as reviewer to this PR to have an additional pair of :eyes: focussing on the "Observable" aspects.

File: pkg/allocator/cache.go (hunk 9 above)
Comment (joestringer) [reply pkg/allocator/cache.go:312]: I see, so the auth gc subscribes to the stream via `job.Observer()` calls that already exist in the tree. The Observer uses the `params.IdentityChanges` `Observable[cache.IdentityChange]` defined in `pkg/auth` (sidenote, which is defined in the Cell Params in order to inform Hive about the dependency). `job.Observer` is a special kind of cell that calls this `Observe()` call once on startup, which triggers this code to spawn a goroutine to prevent blocking, does the listing etc using `mapGC.handleIdentityChange(). Then eventually this function calls into the inner observable `c.changeSrc.Observe()` which spawns its own goroutine to then react to any subsequent updates and in the end that also calls back to the `mapGC.handleIdentityChange()` to implement subsequent changes. (Reading back through your message, I believe I'm just paraphrasing exactly what you said already)

So yeah it seems like we have an extra temporary goroutine per subscription which might not be strictly necessary, but we can always optimize that further in future.

Then the other part to me is that I would have thought that "spawn a goroutine to join the stream, wait for all previous events to be synced, call `next()` for all existing elements, then wait on further events and process them on demand" is a common pattern for Multicast streams. It's a little surprising to me that `pkg/identity` needs to be implementing a nontrivial portion of this logic itself rather than having that as part of the Multicast library (maybe even with part of the async logic built into `job.Observer`?). But then maybe I'm jumping too far ahead already, this is the implemnetation that is testing out the boundaries for the library code vs. the source vs. the sink, and we'll use this implementation to further improve the way this code fits into the rest of the agent.

From that perspective I'm fine with this implementation, even if I do think we could follow up on some further improvements/streamlining of the implementation :+1: Thanks for explaining it to me.

File: pkg/identity/cache/allocator.go (hunk 8 above)
Comment (joestringer) [reply pkg/identity/cache/allocator.go]: Why does this need `complete()` function need to be called asynchronously?

File: pkg/allocator/cache.go (hunk 9 above)
Comment (joestringer) [reply pkg/allocator/cache.go:312]: > One special thing in this use-case is the additional goroutine in pkg/allocator/cache.go. It ensures that the current state of the cache gets re-played at the beginning - to mimic the behaviour of Resource[CiliumIdentity]. But this is a short-lived goroutine which should terminate as soon as the state of the cache has been re-played (or the context has been cancelled in the meantime) and the observation got delegated to the actual Observable mentioned above.

I think it could help to add this sort of commentary above the goroutine. This would help readers to understand how often this goroutine gets created, when the goroutine completes, and assure themselves that it doesn't leak.

File: pkg/allocator/cache.go (hunk 10 above)
Comment (joestringer) [reply pkg/allocator/cache.go:323]: I see, so `c.changeSrc.Observe()` is actually the anonymous `Observe()` function defined in `stream.Multicast[...](...)` in its FuncObservable (via `c.ChangeSrc` function populated in the cache constructor with a Multicast stream creation call), which I agree will spawn its own goroutine to handle subsequent events and unblock this caller + release the readlock.

**Review by joestringer — APPROVED** [review]
Just minor nits remaining: mostly just documenting each `go ...` call to declare why it's necessary and how we know it will complete.

File: pkg/identity/cache/allocator.go [hunk 11]
```diff
@@ -482,3 +486,47 @@ func (m *CachingIdentityAllocator) RemoveRemoteIdentities(name string) {
...
+		// Observe the underlying allocator for changes and map the events to
+		// identities.
+		m.IdentityAllocator.Observe(
+			ctx,
+			func(change allocator.AllocatorChange) {
+				var idLabels labels.Labels = nil
+				if change.Key != nil {
+					idLabels = labels.Labels{}
+					for k, v := range change.Key.GetAsMap() {
+						label := labels.ParseLabel(fmt.Sprintf("%s=%s", k, v))
```
Comment (christarazi) [inline pkg/identity/cache/allocator.go]: Let's avoid the use of `fmt.Sprintf` due to #19571

File: pkg/identity/cache/allocator.go (hunk 8 above)
Comment (christarazi) [reply pkg/identity/cache/allocator.go]: I'm also not clear on that. It would be good to document.

File: pkg/allocator/cache.go (hunk 5 above)
Comment (christarazi) [reply pkg/allocator/cache.go:184]: I believe that above is correct--it's sort of like a "hairpin". Identity is created within one loop and once that creation event is received on the same node that created the identity, we then `emitChange`. I think the location of the latter is correct because we should essentially just mirror what `sendEvent` does.

File: pkg/identity/cache/allocator.go (hunk 11 above)
Comment (mhofstetter) [reply pkg/identity/cache/allocator.go]: thanks for the hint - will change it towards concatenating the values. changed :heavy_check_mark:

File: pkg/identity/cache/allocator.go (hunk 8 above)
Comment (mhofstetter) [reply pkg/identity/cache/allocator.go]: AFAIU, the main purpose is to adhere to the description of the Observable interface and call it asynchronously the same way as it would if the identityAllocator would be set. mainly to prevent faulty assumptions.

-> definitely worth to document the thought :heavy_check_mark:

File: pkg/allocator/cache.go (hunk 9 above)
Comment (mhofstetter) [reply pkg/allocator/cache.go:312]: agree. documented it in the code :heavy_check_mark:

**Comment by mhofstetter** [comment]
@joestringer @christarazi thanks for your input. i addressed your suggestions in the latest force push.

File: pkg/allocator/cache.go (hunk 5 above)
Comment (giorio94) [reply pkg/allocator/cache.go:184]: Thanks! My concern was mainly that, when replaying the state, an event would be emitted also for the entries that have been `inserted` but the corresponding notification has not yet been received from the kvstore. But at this point I guess it should not be a problem given that it will be eventually received (generating a spurious upsert event).

File: pkg/identity/cache/allocator.go [hunk 12]
```diff
@@ -482,3 +486,49 @@ func (m *CachingIdentityAllocator) RemoveRemoteIdentities(name string) {
...
+					Kind:   IdentityChangeKind(change.Kind),
+					ID:     identity.NumericIdentity(change.ID),
+					Labels: idLabels,
+				})
+			},
+			complete,
+		)
+	} else {
+		// Calling complete from a new go routine, the same way as it would happen if the IdentityAllocator would
+		// be set. This is how it should be expected according to stream.Observable and prevents faulty assumptions.
```
Comment (joestringer) [inline pkg/identity/cache/allocator.go]: I'm still confused, bear with me:

> the same way as it would happen if the IdentityAllocator would be set

What is that way? Is there a function reference or other docs description that the reader can compare with? Given that `complete()` is an opaque function, it's hard to understand what `complete` does here and why that means it needs to be run in a goroutine. This comment assumes that the reader already knows this code very well, and then explains to them that "yes it works the way you expect". But for a fresh reader, they will not have the context, so then how will they build up the context to validate this assumption?

> This is how it should be expected according to stream.Observable

`Observable` interface says:

          // Observe a stream of values as long as the given context is valid.                                                                                                                              
          // 'next' is called for each item, and finally 'complete' is called                                                                                                                               
          // when the stream is complete, or an error has occurred.                                                                                                                                         
          //                                                                                                                                                                                                
          // Observable implementations are allowed to call 'next' and 'complete'                                                                                                                           
          // from any goroutine, but never concurrently.  

Nothing in the above tells me that `complete` must be run in a new goroutine.

> prevents faulty assumptions.

What are those faulty assumptions?

File: pkg/allocator/cache.go [hunk 13]
```diff
@@ -274,3 +290,51 @@ func (c *cache) numEntries() int {
...
+	ID   idpool.ID
+	Key  AllocatorKey
+}
+
+// Observe the allocator changes. Conforms to stream.Observable.
+// Replays the current state of the cache when subscribing.
+func (c *cache) Observe(ctx context.Context, next func(AllocatorChange), complete func(error)) {
+	// This short-lived go routine serves the purpose of replaying the current state of the cache before starting
+	// to observe the actual source changeSrc. ChangeSrc is backed by a stream.FuncObservable, that will start its own
+	// go routine. Therefore, the current go routine will stop and free the lock on the mutex after the registration.
```
Comment (joestringer) [inline pkg/allocator/cache.go:311]: :+1: this helps a lot, thanks!

File: pkg/allocator/cache.go (hunk 6 above)
Comment (dylandreimerink) [reply pkg/allocator/cache.go]: I believe `TODO`s are frowned upon, GH issues are preferred

File: pkg/identity/cache/allocator.go [hunk 14]
```diff
@@ -482,3 +486,49 @@ func (m *CachingIdentityAllocator) RemoveRemoteIdentities(name string) {
...
+	Labels labels.Labels
+}
+
+// Observe the identity changes. Conforms to stream.Observable.
+// Replays the current state of the cache when subscribing.
+func (m *CachingIdentityAllocator) Observe(ctx context.Context, next func(IdentityChange), complete func(error)) {
+	if m.IdentityAllocator != nil {
+		// Observe the underlying allocator for changes and map the events to
+		// identities.
+		m.IdentityAllocator.Observe(
```
Comment (dylandreimerink) [inline pkg/identity/cache/allocator.go]: Nit: Consider using the `stream.Map` operator so you can reduce the boilerplate

**Review by dylandreimerink — APPROVED** [review]
Looks good to me overall!

File: pkg/identity/cache/allocator.go (hunk 12 above)
Comment (dylandreimerink) [reply pkg/identity/cache/allocator.go]: There is no specific contract with regards to calling `complete` from the same or a different routine as `Observe`. But looking at existing examples, we always call `complete` from a goroutine spawned inside `Observe`. Due to that behavior, we might that users of `stream.Observables` which for some reason don't expect complete to be called before the call to `Observe` returns. I assume those are the "prevents faulty assumptions." referred to. 

The faulty assumption could lead to the following which would deadlock if `complete` is called synchronously: 
```
mu.Lock()
someStream.Observe(ctx, func(){...}, func(err error){
  mu.Lock()
  defer mu.Unlock()
})
mu.Unlock()
```

So by calling `complete` from a goroutine we avoid deadlocks by uninformed users which can be called "resilience". But its also implementing an implicit contract, so perhaps if we want to allow this we should document it. 

TL;DR calling `complete` without a goroutine is valid as long as users of the API don't use it wrongly

Comment (joestringer) [reply pkg/identity/cache/allocator.go]: @dylandreimerink I see what you're saying, but it seems odd to me that a caller would grab a mutex outside the Observe and also in the complete function that they provide to `Observe()`. I would assume that the functions that are called in `Observe()` can be called from the current thread, so that pattern would just be unsafe. In fact, from the `Observable` documentation, it sounds like Observable implementations *could* call it even from the same goroutine, which says that the locking example above is not respecting the `Observe()` API. ie grabbing a mutex then calling `Observe(...)` where the `Observable` can choose the same goroutine to call `complete()` could easily trigger a deadlock. Therefore we shouldn't assume that people will write such code, we'll catch it in review and reject it because it doesn't use the `Observe()` function correctly.

Comment (joestringer) [reply pkg/identity/cache/allocator.go]: One more bit of background: We should have reasons for goroutines, and avoid spawning goroutines if they're not necessary. They do take a little bit of resources, and they also add complexity due to (1) lifecycle management, ensuring they're completed and (2) by making things more asynchronous. I'd imagine there are good reasons for a lot of the existing `complete()` goroutine cases today, but it makes me nervous to see "well run it in a goroutine to avoid bad assumptions", because we can still make other bad assumptions around the goroutines and then we have bad assumptions _and_ additional async complexity.

Comment (joestringer) [reply pkg/identity/cache/allocator.go]: Maybe... could you rephrase exactly what the concrete problem is with having `complete()` run synchronously? I'm clearly still missing something here.

Comment (joestringer) [reply pkg/identity/cache/allocator.go]: Stepping back through, I think the argument is that since `CachingIdentityAllocator` -> `Allocator` -> `cache` -> `newCache()` initializes the stream with a `stream.Multicast[...]`, and since the implementation of `stream.Multicast` runs `complete()` from a goroutine, this code must also do the same. Broadly that seems like a consistency argument, which seems fine at face value. Based on the `Observable()` interface API, it should be equally fine to just run the completion function directly without the goroutine.

Comment (joestringer) [reply pkg/identity/cache/allocator.go]: Assuming that my latest comment is correct, my suggestion would be either:

(a) drop the goroutine since it's not needed, or
(b) document that assumption more explicitly:

```suggestion
		// Calling complete from a new go routine, the same way as it would happen from stream.Multicast() in the Observe() function above.
```

File: pkg/allocator/cache.go (hunk 6 above)
Comment (mhofstetter) [reply pkg/allocator/cache.go]: i removed the comment. will discuss with @joamaki whether we want to create a GH issue or not

File: pkg/identity/cache/allocator.go (hunk 14 above)
Comment (mhofstetter) [reply pkg/identity/cache/allocator.go]: thanks @dylandreimerink . even though it doesn't remove that much boilerplate, i agree that it's way better to use APIs of package `stream` to unify such logic and the observable-patterns (instead of having to re-invent & document them in the business logic)

File: pkg/identity/cache/allocator.go (hunk 12 above)
Comment (mhofstetter) [reply pkg/identity/cache/allocator.go]: @joestringer @dylandreimerink 

i came up with c) : removing the actual go-routine from the id allocation business logic and using the existing `stream` API `stream.Empty[allocator.AllocatorChange]` in case where `m.IdentityAllocator` is `nil`. I refactored the logic a little bit - so the only difference is the observable (empty or m.IdentityAllocator). IMO this way the code should be readable and understandable from a id allocation point of view - if someone is understanding the observable patterns (there are enough pointers in the go doc of package stream).

even though the called `stream.Empty` is doing the same thing (immediately calling complete from a new go-routine), it's up to the `stream` package to provide the necessary (and allowed) observable patterns as API and document them well (so not every business logic need to document this again and again).

It's really a good discussion and i think there are valid arguments to keep or remove the go-routine. It's just that i think the context to discuss this in this "id allocation business related" PR isn't the right one. It should be part of the `stream` API and is relevant for all usages of it. -> can we discuss this in a follow up? will bring this up once @joamaki is back.

WDYT?

**Comment by mhofstetter** [comment]
@dylandreimerink thanks a lot for your review: i removed the `TODO`, refactored towards using `stream.Map`(:pray:) and `stream.Empty` (discovered after reading through the API of `stream`)

File: pkg/allocator/cache.go (hunk 9 above)
Comment (mhofstetter) [reply pkg/allocator/cache.go:312]: resolved in favor of [follow-up discussion](https://github.com/cilium/cilium/pull/26373#discussion_r1238826898)

**Comment by mhofstetter** [comment]
The latest force push changes that the global identity allocator needs to be initialized before the starting to observe the identity allocator for changes.

This is necessasry, because the initialization of the global identity allocator is performed [asynchronously when the daemon gets started](https://github.com/cilium/cilium/blob/main/pkg/identity/cache/allocator.go#L171-L214).

File: pkg/identity/cache/allocator.go (hunk 12 above)
Comment (mhofstetter) [reply pkg/identity/cache/allocator.go]: next update:

Trying to understand the initial reason to differentiate between whether `m.IdentityAllocator` is initialized or not triggered me to dive into this a little bit more. I came to the conclusion that it's necessary to wrap the whole `(m *CachingIdentityAllocator) Observe` in an additional short lived go-routine (:see_no_evil:) which waits until the `m.IdentityAllocator` is initialized before starting to observe it.

The reason for this is that the global identity allocator is initialized [asynchronously](https://github.com/cilium/cilium/blob/main/pkg/identity/cache/allocator.go#L171-L214) - and might not be properly intialized once starting to observe. This would result in the immediate completion. IMO in these situations its expected that the observation waits until the identity allocator is ready and initialized anyway - otherwise the depending functionality doesn't work as expected (e.g. auth gc job would stop listening for these events)

This comes with the "advantage" that we don't need to discuss whether `complete` is called from a separate go-routine by itself. (because the whole functionality runs in its own go-routine)

I documented this additional short-lived go-routine! (same situation here - locking the mutex (necessary to ensure that the identitAllocator is _really_ set) shouldn't be a problem - because the inner observer starts a go-routine too)

PTAL @joestringer @dylandreimerink :pray: 

PS: yes at best we would start to properly modularize the global identity allocator and register proper lifecycle hooks for the initialization (including the backends) - so dependent components can expect a initialized component.

**Review by jrajahalme — APPROVED** [review]
LGTM

**Comment by mhofstetter** [comment]
rebased to `main` (failed tests related to https://github.com/cilium/cilium/pull/26496)

**Review by christarazi — APPROVED** [review]

File: pkg/identity/cache/allocator.go (hunk 8 above)
Comment (mhofstetter) [reply pkg/identity/cache/allocator.go]: resolved in favor of https://github.com/cilium/cilium/pull/26373#discussion_r1238826898
